nixos/modules/services/security/bitwarden-directory-connector.nix

```diff
@@ -0,0 +1,346 @@
+{
+  pkgs,
+  config,
+  lib,
+  ...
+}:
+with lib; let
+  cfg = config.services.bitwarden_directory_connector;
+
+  ldap_data = ''
+    {
+      "ssl": ${boolToString cfg.ldap.ssl},
+      "startTls": ${boolToString cfg.ldap.startTls},
+      "sslAllowUnauthorized": ${boolToString cfg.ldap.sslAllowUnauthorized},
+      "port": ${toString cfg.ldap.port},
+      "currentUser": false,
+      "ad": ${boolToString cfg.ldap.ad},
+      "pagedSearch": true,
+      "password": "to_be_replaced",
+      "hostname": "${cfg.ldap.hostname}",
+      "rootPath": "${cfg.ldap.root}",
+      "username": "${cfg.ldap.username}"
+    }
+  '';
+
+  sync_data = ''
+    {
+      "removeDisabled": ${boolToString cfg.sync.removeDisabled},
+      "overwriteExisting": ${boolToString cfg.sync.overwriteExisting},
+      "largeImport": ${boolToString cfg.sync.largeImport},
+      "creationDateAttribute": "${cfg.sync.creationDateAttribute}",
+      "memberAttribute": "${cfg.sync.memberAttribute}",
+
+      "useEmailPrefixSuffix": ${boolToString cfg.sync.emailPrefixSuffix.enable},
+      ${optionalString cfg.sync.emailPrefixSuffix.enable ''
+      "emailPrefixAttribute": "${cfg.sync.emailPrefixSuffix.prefixAttribute}",
+      "emailSuffix": "${cfg.sync.emailPrefixSuffix.suffix}",
+    ''}
+
+      "users": ${boolToString cfg.sync.users.enable},
+      ${optionalString cfg.sync.users.enable ''
+      "userPath": "${cfg.sync.users.path}",
+      "userObjectClass": "${cfg.sync.users.objectClass}",
+      "userEmailAttribute": "${cfg.sync.users.emailAttribute}",
+      "userFilter": "${cfg.sync.users.filter}",
+    ''}
+
+      "groups": ${boolToString cfg.sync.groups.enable},
+      ${optionalString cfg.sync.groups.enable ''
+      "groupPath": "${cfg.sync.groups.path}",
+      "groupObjectClass": "${cfg.sync.groups.objectClass}",
+      "groupNameAttribute": "${cfg.sync.groups.nameAttribute}",
+      "groupFilter": "${cfg.sync.groups.filter}",
+    ''}
+
+      "interval": 5
+    }
+  '';
```

Comment:
This all seems like it would be much more robust with `builtins.toJSON`

Reply:
Another functions I learnt about.  
It did make it much easier to work with.